File: Resueltos_por_competidor/Juan_Martinez/NewYearTransportation.cpp
/*
 * Autor: Juan Martinez
 * Problema: New Year Transportation (500A)
 * Juez online: Codeforces
 * Veredicto: Accepted
 * Url: https://codeforces.com/problemset/problem/500/A
 * Difficulty: 1000
 */
#include <bits/stdc++.h>

using namespace std;

vector<vector<int>> adj;
bool visited[30001];

void agregarNodo(int a, int b){
    adj[a].push_back(b);
}

void dfs(int start){
    if(!visited[start]){
        visited[start] = true;
        for(int i : adj[start]){
            dfs(i);
        }
    }
}

int main()
{
    int a, b, temp;
    cin >> a >> b;
    adj.resize((a+1));
    for(int i = 0; i < (a-1); i++) {
        cin >> temp;
        agregarNodo((i+1), temp+(i+1));
    }
    dfs(1);
    if(visited[b]) cout<<"YES"<<endl;
    else cout<<"NO"<<endl;
    return 0;
}

File: Resueltos_por_competidor/Juancho_Garcia/NewYearTransportation.java
/*
 * Autor: Juan Andres Garcia
 * Problema: New Year Transportation
 * Juez online: Codeforces
 * Veredicto: Accepted
 * Url: https://codeforces.com/problemset/problem/500/A
 *  **/

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

public class NewYearTransportation {
	
	
	
	
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		int cantidadCasas = sc.nextInt();
		int casaLlegada = sc.nextInt();
		boolean llego = false;
		Map<Integer, Integer> portales = new HashMap<>();
		
		
		for(int i = 1; i <= cantidadCasas-1; i++) {
			
			int portal = sc.nextInt();
			
			portales.put(i, i + portal);
			
			
		}
		
		
		int casaActual = 1;
		
		
		while(casaActual < cantidadCasas) {
			
			if(casaActual == casaLlegada) {
				llego = true;
				break;
			} 
			if (casaActual > casaLlegada) {
				break;
			}
			
			casaActual = portales.get(casaActual);
			
		}
		
		if (casaActual == casaLlegada) {
			llego = true;
		}
		
		
		if(llego) {
			
			System.out.println("YES");
		}else {
			
			System.out.println("NO");
		}
	
		
		
	}

}
